sql: implement ALTER/CREATE SEQUENCE ... RESTART

nnaka2992 opened the pull request:
fixes #74496

Extended the `ALTER SEQUENCE` command to support restarting 
the sequence with and without value. The new command call 
`nextval('sequence_name')` internally as PostgreSQL does. 
At the same time this supports 
`CREATE SEQUENCE sequence_name RESTART` command for PostgreSQL 
compatibility.

Release note (sql change): Add a new RESTART option to ALTER and
CREATE SEQUENCE which sets the nextval to arbitary number or
a value set for START.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/78958)
<!-- Reviewable:end -->

otan commented on pkg/sql/sem/tree/create.go:
```diff
@@ -1703,6 +1703,19 @@ func (node *SequenceOptions) Format(ctx *FmtCtx) {
+			// TODO(knz): replace all this with ctx.FormatNode if/when
+			// the restart option supports expressions.
+			if ctx.flags.HasFlags(FmtHideConstants) {
+				ctx.WriteByte('0')
```
is there a reason we choose `0` instead of `_`?

nnaka2992 replied:
removed `0` since restart should be able to run without value, thx

otan replied:
we should still use `_`

otan commented on pkg/sql/sem/tree/create.go:
```diff
@@ -1703,6 +1703,19 @@ func (node *SequenceOptions) Format(ctx *FmtCtx) {
+				ctx.WriteString("WITH ")
+			}
+			// TODO(knz): replace all this with ctx.FormatNode if/when
+			// the restart option supports expressions.
```
it doesnt look like restart supports expressions, so i'd remove this TODO

otan replied:
looks like this TODO should still be removed

otan commented on pkg/sql/sem/tree/create.go:
```diff
@@ -1745,6 +1758,9 @@ type SequenceOption struct {
 	ColumnItemVal *ColumnItem
+
+	// Whether the IntVal is optional.
+	OptionalVal bool
```
do we need this? can we use `IntVal == nil` as the check?

otan commented on pkg/sql/catalog/descpb/structured.proto:
```diff
@@ -1062,14 +1064,14 @@ message TableDescriptor {
     // AS option value for CREATE SEQUENCE, which specifies the default
     // min and max values a sequence can take on.
-    optional string as_integer_type = 8 [(gogoproto.nullable) = false];
+    optional string as_integer_type = 9 [(gogoproto.nullable) = false];
```
we should not renumber protobufs fields - otherwise it breaks compatibility with older versions (e.g. older versions expect as_integer_type = 8, but you've changed it to nine so they are broken).
use `restart = 10` instead - the order in which they appear does not matter, and revert the other changes.

otan replied:
this is still an issue

otan reviewed (commented):
this is close, thanks @nnaka2992!

can you also squash your PR into one commit, as blathers tells you to do?

otan reviewed (commented):
looks like the changes didn't get committed? let me know when this is ready for review

nnaka2992 commented:
Sorry for the wrong commit, things has corrupted with git rebase. Now everything is up to date.
I appreciate your patience.

otan reviewed (approved):
thanks!

bors r+

otan reviewed (changes requested):
bors r-

ah there's a failing test

otan commented on pkg/sql/sem/tree/create.go:
```diff
@@ -1703,6 +1703,17 @@ func (node *SequenceOptions) Format(ctx *FmtCtx) {
+				ctx.WriteString("WITH ")
+			}
+			if ctx.flags.HasFlags(FmtHideConstants) {
+				ctx.WriteByte('_')
```
ah! maybe we should use 0 here like everywhere else

nnaka2992 replied:
Fixed

nnaka2992 replied:
I realized that the commit need fix. With `ctx.WriteByte('0')`, restart command with literal removed format, it adds 0.
I will change soon, so please ignore this commit. thanks

nnaka2992 replied:
fixed

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/78958)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nnaka2992 and @otan)

___
*[pkg/sql/alter_sequence.go, line 144 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/78958#-MzvHTHq-eWPkMx9a7C4:-MzvHTHq-eWPkMx9a7C5:b7a0r96) ([raw file](https://github.com/cockroachdb/cockroach/blob/725337c709270527c28e109a8f016b7fa26d30e0/pkg/sql/alter_sequence.go#L144)):*
does this match this behavior? (from the postgres docs)
> In contrast to a setval call, a RESTART operation on a sequence is transactional and blocks concurrent transactions from obtaining numbers from the same sequence. If that's not the desired mode of operation, setval should be used.
___
*[pkg/sql/create_sequence.go, line 163 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/78958#-MzvHK1F48dJEGaedtrM:-MzvHK1GCPCn4sfbPXs0:brv5e9b) ([raw file](https://github.com/cockroachdb/cockroach/blob/725337c709270527c28e109a8f016b7fa26d30e0/pkg/sql/create_sequence.go#L163)):*
> ```Go
> 
> 	var startVal int64
> 	if desc.SequenceOpts.Restart != nil {
> ```

i don't think this is right. RESTART is not valid for CREATE SEQUENCE. it's only to be used with ALTER SEQUENCE
___
*[pkg/sql/parser/sql.y, line 8118 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/78958#-MzvHch5ED5dmX0e3HCS:-MzvHch5ED5dmX0e3HCT:b70y7b3) ([raw file](https://github.com/cockroachdb/cockroach/blob/725337c709270527c28e109a8f016b7fa26d30e0/pkg/sql/parser/sql.y#L8118)):*
> ```Yacc
> | START WITH signed_iconst64   { x := $3.int64()
>                                  $$.val = tree.SequenceOption{Name: tree.SeqOptStart, IntVal: &x, OptionalWord: true} }
> | RESTART                      { $$.val = tree.SequenceOption{Name: tree.SeqOptRestart} }
> ```

RESTART is not allowed for CREATE SEQUENCE, so i think we should have a different grammar rule for sequence options that are only allowed in ALTER SEQUENCE

also, the tests should verify that using RESTART in CREATE SEQUENCE is a parsing error


<!-- Sent from Reviewable.io -->

nnaka2992 commented:
I thought this is weird but the current PostgreSQL implementation support `RESTART` option with `CREATE SEQUENCE`.
The following queries run on PostgreSQL 14.2

```
psql (14.2 (Debian 14.2-1.pgdg110+1))
Type "help" for help.

admin=# SELECT version();
                                                           version                                                           
-----------------------------------------------------------------------------------------------------------------------------
 PostgreSQL 14.2 (Debian 14.2-1.pgdg110+1) on x86_64-pc-linux-gnu, compiled by gcc (Debian 10.2.1-6) 10.2.1 20210110, 64-bit
(1 row)

admin=# CREATE SEQUENCE test RESTART 10;
CREATE SEQUENCE
admin=# select nextval('test');
 nextval 
---------
      10
(1 row)

```

As PostgreSQL compatible, `CREATE SEQUENCE sequence_name RESTART` should be supported.

nnaka2992 commented:
Of course I can omit  `CREATE SEQUENCE RESTART` if cockroachDB choose not to support these!

rafiss commented:
hm, you're right, I guess Postgres supports it even though it's not documented on: https://www.postgresql.org/docs/14/sql-createsequence.html

I'm still curious about my other question. We should match this behavior:
> In contrast to a setval call, a RESTART operation on a sequence is transactional and blocks concurrent transactions from obtaining numbers from the same sequence. If that's not the desired mode of operation, setval should be used.

rafiss commented:
Also, I am all right with allowing RESTART in CREATE SEQUENCE. Can you add a test for that, so the behavior is verified?

nnaka2992 commented:
Sorry, I did something wrong on git rebase.
Please ignore the commit

nnaka2992 committed:
sql: implement ALTER/CREATE SEQUENCE ... RESTART

fixes cockroachdb#74496
what you did about it:
Extended the ALTER SEQUENCE command to support restarting
the sequence with and without value. The new command call
nextval('sequence_name') internally as PostgreSQL does.
At the same time this supports
CREATE SEQUENCE sequence_name RESTART command for PostgreSQL
compatibility.

Release note (sql change): Add a new RESTART option to ALTER and
CREATE SEQUENCE which sets the nextval to arbitary number or
a value for START.

nnaka2992 commented:
> hm, you're right, I guess Postgres supports it even though it's not documented on: https://www.postgresql.org/docs/14/sql-createsequence.html
> 
> I'm still curious about my other question. We should match this behavior:
> 
> > In contrast to a setval call, a RESTART operation on a sequence is transactional and blocks concurrent transactions from obtaining numbers from the same sequence. If that's not the desired mode of operation, setval should be used.

I changed some code to use transaction when applying restart is in transaction.

nnaka2992 commented:
> Also, I am all right with allowing RESTART in CREATE SEQUENCE. Can you add a test for that, so the behavior is verified?

I added test code for CREATE SEQUENCE RESTART

rafiss reviewed (changes requested):
we still need tests for `CREATE SEQUENCE RESTART` either in pkg/sql/logictest/testdata/logic_test/sequences or pkg/sql/logictest/testdata/logic_test/alter_sequence

> Reviewed 4 of 12 files at r8, 2 of 2 files at r10, 3 of 6 files at r11, 2 of 915 files at r13, 1 of 1 files at r14, 913 of 913 files at r15, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/78958#-:-N06qSw_Be-kSKPHkD5w:b7bcslm)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nnaka2992 and @otan)

___
*[pkg/sql/alter_sequence.go, line 144 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/78958#-N06oUn_EtA64ND6k5Rl:-N06oUn_EtA64ND6k5Rm:bzfn64o) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3f6267cced076569e84f030a0d9b5cef585b29b/pkg/sql/alter_sequence.go#L144)):*
> ```Go
> 	}
> 	if opts.Restart != nil {
> 		if err := params.p.createdSequences.addCreatedSequence(desc.ID); err != nil {
> ```

i don't think it's correct to add this to "created sequences" here
___
*[pkg/sql/sequence.go, line 306 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/78958#-N06plNG8ISo9xGV4NI9:-N06plNG8ISo9xGV4NIA:b7lpv97) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3f6267cced076569e84f030a0d9b5cef585b29b/pkg/sql/sequence.go#L306)):*
> ```Go
> // setSequenceValueHelper is shared by SetSequenceValue and SetSequenceValueByID
> // to set the given sequence to a new given value.
> func setSequenceValueHelper(
> ```

i think the rebase needs to be redone. this `setSequenceValueHelper` function has been removed
___
*[pkg/sql/sequence.go, line 334 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/78958#-N06prKhEaCDfGZ2GKvW:-N06prKhEaCDfGZ2GKvX:bee5fe) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3f6267cced076569e84f030a0d9b5cef585b29b/pkg/sql/sequence.go#L334)):*
> ```Go
> 
> 	createdInCurrentTxn := p.createdSequences.isCreatedSequence(descriptor.GetID())
> 	if createdInCurrentTxn {
> ```

i don't think this is correct for RESTART. according to the docs i shared, `ALTER SEQUENCE ... RESTART` _always_ should use `p.txn`


<!-- Sent from Reviewable.io -->
